app-text/mandoc/mandoc-1.14.3.recipe

```diff
@@ -0,0 +1,113 @@
+SUMMARY="A suite of tools compiling mdoc from the OpenBSD project"
+DESCRIPTION="A suite of tools compiling mdoc, the roff macro language of \
+choice for BSD manual pages, and man, the predominant historical language for \
+UNIX manuals. It is small, ISO C, ISC-licensed, and quite fast. The main \
+component of the toolset is the mandoc utility program, based on the libmandoc \
+validating compiler, to format output for UTF-8 and ASCII UNIX terminals, \
+HTML 5, PostScript, and PDF."
+HOMEPAGE="http://mandoc.bsd.lv/"
+# Copyrights should be updated on every new release
+# http://mandoc.bsd.lv/cgi-bin/cvsweb/LICENSE
+COPYRIGHT="2008-2012, 2014 Kristaps Dzonsons
+	2010-2017 Ingo Schwarze
+	2009, 2010, 2011, 2012 Joerg Sonnenberger
+	2013 Franco Fichtner
+	2014 Baptiste Daroussin
+	2016 Ed Maste
+	2017 Michael Stapelberg
+	1999, 2004 Marc Espie
+	1998, 2004, 2010 Todd C. Miller
+	2008, 2017 Otto Moerbeek
+	2004 Ted Unangst
+	1994 Christos Zoulas
+	2003, 2007, 2008, 2014 Jason McIntyre
+	2014 Baptiste Daroussin
+	1989, 1990, 1993, 1994 The Regents of the University of California
+	1994 Christos Zoulas"
+LICENSE="ISC
+	BSD (2-clause)
+	BSD (3-clause)"
+REVISION="1"
+SOURCE_URI="http://mandoc.bsd.lv/snapshots/mandoc-$portVersion.tar.gz"
+CHECKSUM_SHA256="0b0c8f67958c1569ead4b690680c337984b879dfd2ad4648d96924332fd99528"
+PATCHES="mandoc-$portVersion.patchset"
+
+ARCHITECTURES="!x86_gcc2 x86_64"
+SECONDARY_ARCHITECTURES="x86"
+
+PROVIDES="
+	mandoc$secondaryArchSuffix = $portVersion
+	cmd:apropos = $portVersion
+	cmd:demandoc = $portVersion
+	cmd:makewhatis = $portVersion
+	cmd:man = $portVersion
+	cmd:mandoc = $portVersion
+	cmd:mandocd = $portVersion
+	cmd:soelim = $portVersion
+	cmd:whatis = $portVersion
+	"
+REQUIRES="
+	haiku$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	"
+CONFLICTS="
+	cmd:man
+	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libz$secondaryArchSuffix
+	"
+BUILD_PREREQUIRES="
+	cmd:cc$secondaryArchSuffix
+	cmd:make
+	"
+
+TEST_REQUIRES="
+	cmd:diff
+	cmd:perl
+	"
+BUILD_PREREQUIRES+="$TEST_REQUIRES"
+
+USER_SETTINGS_FILES="settings/man.conf"
+
+BUILD()
+{
+	# Search path
+	local manPathDef=(
+		"$(finddir B_USER_NONPACKAGED_DOCUMENTATION_DIRECTORY)/man"
+		"$(finddir B_USER_DOCUMENTATION_DIRECTORY)/man"
+		"$(finddir B_SYSTEM_NONPACKAGED_DOCUMENTATION_DIRECTORY)/man"
+		"$(finddir B_SYSTEM_DOCUMENTATION_DIRECTORY)/man"
+		)
+	# Cross ref validation path
+	local manPathLint=(
+		"$(finddir B_USER_DOCUMENTATION_DIRECTORY)/man"
+		"$(finddir B_SYSTEM_DOCUMENTATION_DIRECTORY)/man"
+		)
+	cat > configure.local << EOF
+		CFLAGS='-O2 -pipe -D_BSD_SOURCE'
+		DBDIR='$(finddir B_SYSTEM_VAR_DIRECTORY)/mandoc'
+		LN='ln -sf'
+		MANDIR='$manDir'
+		MANPATH_BASE='$(IFS=:; echo "${manPathLint[*]}")'
+		MANPATH_DEFAULT='$(IFS=:; echo "${manPathDef[*]}")'
+		PREFIX='$prefix'
+		SBINDIR="\${PREFIX}/bin"
+		SYSCONFDIR='$portPackageLinksDir/.settings'
+EOF
+	./configure
+
+	make $jobArgs
+}
+
+INSTALL()
+{
+	make install
+}
```

Comment:
Some suggested polishing:
```diff
 INSTALL()
 {
 	make install
+
+	# man.conf is not in /etc/ but in $settingsDir
+	sed -i -e "s|/etc\(/man\.conf\)$|`finddir
+			B_SYSTEM_SETTINGS_DIRECTORY`\1|;" \
+		$manDir/man1/apropos.1 \
+		$manDir/man1/man.1 \
+		$manDir/man5/man.conf.5 \
+		$manDir/man8/makewhatis.8
 }
```
(Another sed command is comming soon.)

Reply:
Since the recipe adds a default `MANPATH` in `configure.local`: `MANPATH_DEFAULT='/boot/home/config/non-packaged/documentation/man:/boot/home/config/documentation/man:/boot/system/non-packaged/documentation/man:/boot/system/documentation/man'`
you might wish to use the following sed:
```
	sed -i \
		-e "/^manpath \/usr\/share\/man/ i manpath `finddir \
			B_USER_NONPACKAGED_DOCUMENTATION_DIRECTORY`/man" \
		-e "/^manpath/ s|/usr/share/man|`finddir \
			B_USER_DOCUMENTATION_DIRECTORY`/man|;" \
		-e "/^manpath/ s|/usr/X11R6/man|`finddir \
			B_SYSTEM_NONPACKAGED_DOCUMENTATION_DIRECTORY`/man|;" \
		-e "/^manpath/ s|/usr/local/man|`finddir \
			B_SYSTEM_DOCUMENTATION_DIRECTORY`/man|;" \
		$manDir/man5/man.conf.5
```
to have these lines:
```
manpath /usr/share/man
manpath /usr/X11R6/man
manpath /usr/local/man
```
replaced by these ones:
```
manpath /boot/home/config/non-packaged/documentation/man
manpath /boot/home/config/documentation/man
manpath /boot/system/non-packaged/documentation/man
manpath /boot/system/documentation/man
```

BTW, the constants associated to the above paths (without the "/man" suffix) are respectively, as we know:
```
B_USER_NONPACKAGED_DOCUMENTATION_DIRECTORY
B_USER_DOCUMENTATION_DIRECTORY
B_SYSTEM_NONPACKAGED_DOCUMENTATION_DIRECTORY
B_SYSTEM_DOCUMENTATION_DIRECTORY
```

BTW, the recipe is hardcoding build-time defaults. That's OK for me, but if you would like to try some changes to have a dynamic default config by calling `find_directory()` when there is no `/system/settings/man.conf`, then this would be even better :)
Well, I guess this would be quite some extra work.

Reply:
@fbrosson There's no `/system/settings/man.conf` by default :laughing:

I would like to use `find_directory()`, but the amount of changes that has to be made is huge, and I don't think it's maintainable. Still, I'd have to figure it out somehow, since this would be a problem once we have proper multi-user support.

Ideas:
- Add a script that would automatically populate `man.conf` with `finddir` output (this would fail once multi-user support exists)
- Find a way to inject `find_directory()` to `mandoc`

Reply:
Yep, I know. mandoc is smart enough to work without any config file :)
(But I did check that the config file was being read (and taken into account) when MANPATH is not defined. My conclusion is that everything is working as expected.)

BTW, there is a small issue: the man package that is currently pre-installed in the official Haiku images does come with a `/system/settings/man.conf`, and the sad news is that the config keywords are all upper-case (`MANPATH`), while mandoc requires it to be lower-case.
But I think it is OK to keep mandoc unchanged. The manual page is very clear about the keywords being lower case. Anyone willing to add custom manpaths can do so with the correct syntax.

Regarding the script to populate a default `man.conf`, I don't think it would be a good idea. The default built-in values are OK. And there is no urgency in teaching mandoc to use dynamic defaults. And, as you pointed, such a change would not be easy to maintain.
So I'd say that the recipe will be OK after you add the sed command to replace `/etc/man.conf` by `/boot/system/settings/man.conf` in the man pages (using the find_dir command in the recipe, of course.)
Regarding the complicated sed command to patch `$manDir/man5/man.conf.5`, it's just a nice to have and I'm OK with keeping `manpath /usr/share/man` + `manpath /usr/X11R6/man` + `manpath /usr/local/man` in that man page.

Regarding the _injection of `find_directory()` in mandoc_, I guess what you mean is finding a smart patch to do that in a way that would be easy to maintain. I'd say that this would be welcome but you really don't need to worry about that.

Reply:
Good news, the manpath code is simple enough that it would be easy to add dynamic folder searching.

Bad news, the code is based on the assumption that it won't fail, but `find_directory()` can. I would need some advice on what to do in case `find_directory()` fails.

Reply:
If it fails, you can still return an hardcoded path.

Reply:
@pulkomandy, I'm a bit surprised by your answer :)
Wouldn't it be better to just skip `find_directory`'s result if it fails?
If `find_directory` ever fails, it will certainly be for a good reason. So returning a hardcoded path in that case would probably not be very useful.
And in case a directory did exist but `find_directory` would fail to return that directory's path, then it would be much easier to detect this error if mandoc were not told to fallback to a hardcoded path.

Reply:
> the code is based on the assumption that it won't fail

We could add error handling support, but I would say that it is not up to us. If the upstream decided to not handle errors at this level, the simplest thing to do is to always return some path, then let them handle errors in appropriate places.

If I was writing my own software, I would handle find_directory failure differently. But here we are doing a port, and I think patching in extra error handling where the original author did not think it was useful is not necessary. Especially as find_directory currently cannot fail, even if the API allows it.

Reply:
mandoc could be patched to either:
* skip any directory that `find_directory()` fails to return,
* or provide a hardcoded fallback.

Now, since we know that `find_directory()` cannot (currently) fail, I suggest that mandoc silently skips any failure, if that ever happens in the future. There is no need, imho, to embed any hardcoded path.
Testing `find_directory()`s return code is the minimum, and this doesn't cost much if mandoc chooses to skip a manpath in case `find_directory()` ever fails.

That said, if it is easier to provide a hardcoded path than ignoring a failed call, then I'm OK with embedding hardcoded paths.